R/chunk_req_perform.R

```diff
@@ -0,0 +1,204 @@
+#' requests
+#'
+#' # ... which you can perform with `req_perform()`
+#' \dontrun{
+#' n <- length(requests)
+#' paths <- character(n)
+#' responses <- list()
+#' for (i in seq2(1, n)) {
+#'   paths[[i]] <- tempfile()
+#'   responses[[i]] <- req_perform(requests[[i]], path = paths[[i]])
+#' }
+#' }
+req_chunk <- function(req,
+                      data,
+                      chunk_size,
+                      apply_chunk) {
+  check_request(req)
+  chunks <- vec_chop_by_size(data, chunk_size)
+  check_function2(apply_chunk, args = c("req", "chunk"))
+
+  n <- length(chunks)
+  requests <- lapply(chunks, function(chunk) apply_chunk(req, chunk))
+  new_chunked_request(requests, chunks)
+}
+
+#' @export
+#' @rdname req_chunk
+#' @inheritParams paginate_req_perform
+#' @param parse_resp A function with one argument `resp` that parses the
+#'   response.
+#' @param progress Whether to show a progress bar. Use `TRUE` to turn on a basic
+#'   progress bar, use a string to give it a name, or see progress_bars for more
+#'   details.
+chunk_req_perform <- function(req,
+                              data,
+                              chunk_size,
+                              apply_chunk,
+                              parse_resp = NULL,
+                              progress = TRUE,
+                              error_call = current_env()) {
+  chunked_requests <- req_chunk(
+    req = req,
+    data = data,
+    chunk_size = chunk_size,
+    apply_chunk = apply_chunk
+  )
+  requests <- chunked_requests$requests
+  the$last_chunks <- chunked_requests$chunks
+
+  parse_resp <- parse_resp %||% function(resp) resp
+  check_function2(parse_resp, args = "resp", call = error_call)
+
+  n <- length(requests)
+  the$last_chunked_responses <- vector("list", n)
+
+  pb <- create_progress_bar(total = n, name = "Request chunks", progress)
+  show_progress <- !is.null(pb)
+  env <- current_env()
+
+  for (i in seq2(1, n)) {
+    req_i <- requests[[i]]
+    try_fetch(
+      resp_i <- req_perform(req_i),
+      error = function(cnd) {
+        cli::cli_abort(
+          "When requesting chunk {i}.",
+          parent = cnd,
+          .envir = env,
+          call = error_call
+        )
+      }
+    )
+    try_fetch(
+      parsed <- parse_resp(resp_i),
+      error = function(cnd) {
+        cli::cli_abort(
+          "When parsing response {i}.",
+          parent = cnd,
+          .envir = env,
+          call = error_call
+        )
+      }
+    )
+
+    the$last_chunked_responses[[i]] <- parsed
+    the$last_chunk_idx <- i
+
+    if (show_progress) cli::cli_progress_update()
+  }
+  if (show_progress) cli::cli_progress_done()
+
+  the$last_chunked_responses
+}
+
+#' @export
+#' @rdname last_response
+last_chunked_responses <- function() {
+  the$last_chunked_responses[seq2(1, the$last_chunk_idx)]
+}
+
+#' @export
+#' @rdname last_response
+last_chunk <- function() {
+  the$last_chunks[[the$last_chunk_idx]]
+}
+
+vec_chop_by_size <- function(x,
```

Comment:
It feels a bit weird to bring in vctrs just for this function. But OTOH it would be a lot of work to accurately recreate the code in this package.